pkg/oss/coco.go

```diff
@@ -0,0 +1,137 @@
+package oss
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"github.com/kubernetes-sigs/alibaba-cloud-csi-driver/pkg/utils/kata/directvolume"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	optDirectAssigned  = "direct"
+	optAnnotations     = "annotations"
+	optEncrypted       = "encrypted"
+	optEncPasswd       = "encPasswd"
+	optKmsKeyId        = "kmsKeyId"
+	fsTypeSecureMount  = "secure_mount"
+	volumeTypeOSS      = "alibaba-cloud-oss"
+	sealedSecretPrefix = "sealed."
+)
+
+func (ns *nodeServer) publishDirectVolume(ctx context.Context, req *csi.NodePublishVolumeRequest, opt *Options) (*csi.NodePublishVolumeResponse, error) {
+	logger := log.WithFields(map[string]interface{}{
+		"NodeServer": "OSSNodeServer",
+		"VolumeId":   req.VolumeId,
+		"TargetPath": req.TargetPath,
+	})
+
+	volumePath := req.TargetPath
+	if isDirectVolumePath(volumePath) {
+		logger.Infof("NodePublishVolume: The mount info for DirectVolume is already exist: %s", volumePath)
+		return &csi.NodePublishVolumeResponse{}, nil
+	}
+
+	device := req.TargetPath
+	volumeType := volumeTypeOSS
+	fsType := fsTypeSecureMount
+	var encrypted, encPasswd, kmsKeyId string
+	annotationsObj := map[string]string{}
+
+	if v := req.Secrets[optEncPasswd]; v != "" {
+		encPasswd = v
+	}
+	for key, value := range req.VolumeContext {
+		switch key {
+		case optEncrypted:
+			encrypted = strings.TrimSpace(value)
+		case optEncPasswd:
+			encPasswd = strings.TrimSpace(value)
+		case optKmsKeyId:
+			kmsKeyId = strings.TrimSpace(value)
+		case optAnnotations:
+			json.Unmarshal([]byte(value), &annotationsObj)
+		}
+	}
+
+	if annotationsObj["kata_fs_type"] != "" {
+		fsType = annotationsObj["kata_fs_type"]
+	}
+	if annotationsObj["kata_volume_type"] != "" {
+		volumeType = annotationsObj["kata_volume_type"]
+	}
+	if annotationsObj["kata_device"] != "" {
+		device = annotationsObj["kata_device"]
+	}
+	annotations, _ := json.Marshal(annotationsObj)
+	annotationsStr := string(annotations)
+	if len(annotations) == 0 || annotationsStr == "null" || annotationsStr == "" {
+		annotationsStr = "{}"
+	}
+	metadata := map[string]string{
+		"bucket":      opt.Bucket,
+		"url":         opt.URL,
+		"otherOpts":   opt.OtherOpts,
+		"path":        opt.Path,
+		"encrypted":   encrypted,
+		"kmsKeyId":    kmsKeyId,
+		"annotations": annotationsStr,
+		"volumeId":    req.GetVolumeId(),
+		"readonly":    fmt.Sprintf("%v", req.Readonly),
+		"targetPath":  req.GetTargetPath(),
+	}
+	if opt.AkSecret != "" && strings.HasPrefix(opt.AkSecret, sealedSecretPrefix) {
+		metadata[AkID] = opt.AkID
+		metadata[AkSecret] = opt.AkSecret
+	}
+	if encPasswd != "" && strings.HasPrefix(encPasswd, sealedSecretPrefix) {
+		metadata[optEncPasswd] = encPasswd
+	}
+
+	mountInfo := directvolume.MountInfo{
+		VolumeType: volumeType,
+		Device:     device,
+		FsType:     fsType,
+		Metadata:   metadata,
+		Options:    strings.Fields(opt.OtherOpts),
```

Comment:
by using `strings.Fields()`, we dropped the support for space in the options, e.g.: `-o key="hello world"` will not work.

Reply:
I don't think the ossfs have such cases.

Reply:
Maybe. But this breaks the current interface. We are parsing the options with shell now, e.g., `-o 'key'` is parsed as `["-o", "key"]`, not `["-o", "'key'"]`